### PR TITLE
Add RFC-0002 (Layer Architecture)

### PR DESCRIPTION
**Pull request by twelho**
This RFC shall describe the requirements of the individual layers of Racklet in more detail. It doesn't describe the implementation details of each layer, those will be covered by upcoming RFCs.

[hackmd-github-sync-badge](https://hackmd.io/HYXkw0gMTSia7-Da14GLpQ)

### CODE REVIEW

File: docs/rfcs/0002-layer-architecture.md [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+
+**Initial Pull Request**: [racklet/racklet#20](https://github.com/racklet/racklet/pull/20)
+
+**Tracking Issue**: [racklet/racklet#21](https://github.com/racklet/racklet/issues/21)
+
+## Summary
+
+_One paragraph explanation of the feature._
+
+This RFC describes the overall Racklet architecture, its defining layers, and requirements for each such layer, derived from [RFC-0001]. For each layer the defining components are described at a high level (avoiding implementation details). The compnoents are associated with their role and five highlighted key requirements from the values and user goals of [RFC-0001].
```
Comment (luxas) [inline docs/rfcs/0002-layer-architecture.md:34]: nit: `compnoents`

File: docs/rfcs/0002-layer-architecture.md [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+
+**TODO**: Embed the graphics into this document when they're ready.
+
+### Goals
+
+_What is in scope for this work?_
+
+- Define well-known layers of Racklet.
+- Describe the requirements for each layer.
+- Describe roughly how to be "Racklet conformant" and what the differences are between Racklet and other similar alternatives.
```
Comment (luxas) [inline docs/rfcs/0002-layer-architecture.md:52]: If we're not answering this now, we can just omit this. I don't think we need to solve this problem in full now anyways.

File: docs/rfcs/0002-layer-architecture.md [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+_This is the technical portion of the RFC. Explain the design in sufficient detail that:_
+
+- _Its interaction with other features is clear._
+- _It is reasonably clear how the feature would be implemented._
+- _Corner cases are dissected by example._
+
+_The section should return to the examples given in the guide-level explanation below, and explain more fully how the detailed proposal makes those examples work._
+
+Racklet is divided into 5 distinct layers, from lowest-level to highest-level:
```
Comment (luxas) [inline docs/rfcs/0002-layer-architecture.md:71]: reverse this so it's in line with how the RFC is read?

File: docs/rfcs/0002-layer-architecture.md [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+**Goals**:
+- Enable the user to observe and manage a Racklet cluster
+- Enable easy deployment of container/VM workloads
+- (Optionally) make a Kubernetes cluster accessible for the user
+
+**Layer components**:
+
+| Component                     | Role                                             | Key Requirements                                                                                        |
+| -------------------------------- | ------------------------------------------------ | ------------------------------------------------------------------------------------------------------- |
+| **VM image building automation** | Define and run VMs declaratively                 | [Improve status quo], [Openness], [Declarative management], [Documentation], [Fast reconfiguration]     |
```
Comment (luxas) [inline docs/rfcs/0002-layer-architecture.md:96]: s/VM image building automation/Micro-Virtual Machine orchestration/

File: docs/rfcs/0002-layer-architecture.md [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+| **Kubernetes VM image**          | Consume/use a Kubernetes cluster                 | [De-facto standards], [Declarative management], [Loose coupling], [Upgradability], [Utilize Kubernetes] |
```
Comment (luxas) [inline docs/rfcs/0002-layer-architecture.md:97]: I'd omit "VM image" here as an implementation detail, and describe it instead as a system of components, which it is gonna be at the end of the day, so e.g. `Kubernetes deployment automation`

File: docs/rfcs/0002-layer-architecture.md [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+| **RMC firmware**                         | Rack hardware control and observability, e.g. fans                           | [Openness], [Declarative management], [Loose coupling], [Observability], [Secure updates]                               |
+
+*[BMC]: Baseboard Management Controller
+*[RMC]: Rack Management Controller
+
+[u-root]: https://u-root.org/
+
+### 2. Electrical layer
+
+**Summary**: The electrical layer backs the power delivery and physical networking requirements of the compute and provides a means to run the firmware on the BMC and RMC (microcontrollers).
```
Comment (luxas) [inline docs/rfcs/0002-layer-architecture.md:147]: maybe spell out RMC and BMC at some point?

File: docs/rfcs/0002-layer-architecture.md [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+### Guide-level explanation
+
+Explain the proposal as if it was already a feature of the project and this would be the documentation for that feature.
+
+- Introducing new named concepts.
+- Explaining the feature largely in terms of examples.
+- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+- If applicable, describe the differences between teaching this to a Racklet administrator versus a Racklet end user.
+
+**TODO**: Mention what is the difference between "reference" implementation and "community" implementations. TBD
```
Comment (luxas) [inline docs/rfcs/0002-layer-architecture.md:237]: Let's omit this in this RFC

File: docs/rfcs/0002-layer-architecture.md [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+[Zero-trust network boot]:    0001-high-level-architecture.html#user-goal-zero-trust-networking
+[End-to-end encryption]:      0001-high-level-architecture.html#user-goal-end-to-end-encryption
+[Physical portability]:       0001-high-level-architecture.html#user-goal-physical-portability
+[Commodity power and I/O]:    0001-high-level-architecture.html#user-goal-commodity-power-io
+
+### Guide-level explanation
+
+Explain the proposal as if it was already a feature of the project and this would be the documentation for that feature.
+
+- Introducing new named concepts.
```
Comment (luxas) [inline docs/rfcs/0002-layer-architecture.md:232]: BMC and RMC would be two examples.

File: docs/rfcs/0002-layer-architecture.md [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+
+[Loose coupling] plays a very important role in the architecture presented here. Racklet could have been designed as a fully integrated system with implementations that are strictly defined by the project, but while this potentially could make the system more compact and simple, it also faces many drawbacks that make it incompatible with the values and goals of the project. For example, Racklet relies heavily on various different projects in the Open Firmware and Cloud Native ecosystems, many of which evolve quickly and provide alternative implementations complying to standard APIs. We want Racklet to be accessible, transparent and modular, which means supporting a wide variety of hardware, and enabling user customization to a great extent. If loose coupling is implemented properly, we believe that the standardized architecture presented here will be relatively simple to maintain and extend, and community-built Racklet solutions will also be able to use the modules and different software implementations effortlessly. In summary, to fulfill the values defined in [RFC-0001] and to avoid ecosystem fragmentation the Racklet project aims to provide interfaces, not implementations.
```
Comment (luxas) [inline docs/rfcs/0002-layer-architecture.md:268]: nit: Open Source Firmware

File: docs/rfcs/0002-layer-architecture.md [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+
+TODO: Explain what is the difference between this try and prior RPi clusters.
+
+At the time of Racklet creation the history of Raspberry Pi (and other SBC) based cluster computers is already very rich. Various private persons, educational insistutes and companies have come up with a wide variety of designs (**TODO: Examples**) for different use cases for at least the past 10 years. What sets Racklet apart from these mostly one-off implementations is it's **specification**. Instead of deriving a specification from some implementation, Racklet as a system is *primarily* defined as a set of RFC documents. This specification is intended to define a **standardized** way to build a miniature compute cluster, from the lowest-level hardware details up to a state-of-the-art software stack. Since the specification is defined from the ground up, we prioritize basing it on the most _secure_ and _modern_ technologies available today, essentially merging the core concepts of prior SBC cluster computer implementations with the state of the art security and fleet management models of large-scale cloud providers.
```
Comment (luxas) [inline docs/rfcs/0002-layer-architecture.md:284]: Examples: https://github.com/kubecloud/thesis

Comment (luxas) [reply docs/rfcs/0002-layer-architecture.md:284]: I'd say past 8 years :)

File: docs/rfcs/0002-layer-architecture.md [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+
+_One paragraph explanation of the feature._
+
+This RFC describes the overall Racklet architecture, its defining layers, and requirements for each such layer, derived from [RFC-0001]. For each layer the defining components are described at a high level (avoiding implementation details). The compnoents are associated with their role and five highlighted key requirements from the values and user goals of [RFC-0001].
+
+[RFC-0001]: 0001-high-level-architecture.md
+
+## Motivation
+
+_Why are we doing this? What use cases does it support? What is the expected outcome?_
```
Comment (luxas) [inline docs/rfcs/0002-layer-architecture.md:40]: let's remove these now. They make the document a bit hard to review.

**Review by luxas — APPROVED** [review]
Thanks for making this, looks great :100:
Left a couple of comments/nits, but overall LGTM

File: docs/rfcs/0002-layer-architecture.md (hunk 2 above)
Comment (twelho) [reply docs/rfcs/0002-layer-architecture.md:52]: RFC-0002 is intended to partially assess that topic, but I'll see if I can reword that to sound less like this document is going to give an in-depth answer.

File: docs/rfcs/0002-layer-architecture.md (hunk 3 above)
Comment (twelho) [reply docs/rfcs/0002-layer-architecture.md:71]: Yep, Markdown was being a bit difficult with reversed lists, but I managed to find a workaround.

File: docs/rfcs/0002-layer-architecture.md (hunk 6 above)
Comment (twelho) [reply docs/rfcs/0002-layer-architecture.md:147]: Mouse over them :)

File: docs/rfcs/0002-layer-architecture.md (hunk 8 above)
Comment (twelho) [reply docs/rfcs/0002-layer-architecture.md:232]: This section needs a bit of a rewrite.

Comment (twelho) [reply docs/rfcs/0002-layer-architecture.md]: Done in https://github.com/racklet/racklet/pull/20/commits/177c36949511bed9eac2607fbc8b39cad96605c1.

File: docs/rfcs/0002-layer-architecture.md (hunk 2 above)
Comment (twelho) [reply docs/rfcs/0002-layer-architecture.md]: Done.

**Comment by twelho** [comment]
I've now overhauled the guide-level explanation and done a bunch of improvements all around. Perhaps a bit embarrassingly, the core of Racklet, the compute unit, was missing in the layer description altogether, so that's now been added as well.

**Review by luxas — APPROVED** [review]
Let's get this in :100:. Sorry I had missed to re-review this.

**Pull request merged**